lib/Models/CatalogItem.js

```diff
@@ -1167,4 +1283,104 @@ function isShownChanged(catalogItem) {
+            if (!JulianDate.equals(terriaClock.currentTime, catalogItem.clock.currentTime)) {
+                updateCurrentTime(catalogItem, terriaClock.currentTime);
+            }
+        });
+
+        updateCurrentTime(catalogItem, catalogItem.terria.clock.currentTime);
+    } else if (!catalogItem._mirroringTerriaClock && defined(catalogItem._removeTerriaClockWatch)) {
+        // We are no longer using the terria clock, remove the subscription.
+        catalogItem._removeTerriaClockWatch();
+        catalogItem._removeTerriaClockWatch = undefined;
+    }
+}
+
+/**
+ * Updates the item.clock.currentTime using the value provided.
+ * This function does this so that all clients subscribed to the items clock with DataSourceClock.definitionChanged are notified.
+ *
+ * @param catalogItem The CatalogItem to set the current time for.
+ * @param updatedTime The new current time to use.
+ * @private
+ */
+function updateCurrentTime(catalogItem, updatedTime) {
+    if (defined(catalogItem.clock)) {
+        catalogItem.clock.currentTime = JulianDate.clone(updatedTime);
+    }
+}
+
+/**
+ * Syncs the time between the terria.clock and the catalogItem.clock.
+ *
+ * The sync is performed in the direction that is required depending on the state change true->false / false->true of useOwnClock.
+ *
+ * @param catalogItem The CatalogItem to sync the time for.
+ * @private
+ */
+function useOwnClockChanged(catalogItem) {
+    // If we are changing the state, copy the time from the clock that was in use to the clock that will be in use.
+    if (catalogItem._lastUseOwnClock !== catalogItem.useOwnClock) {
+        if (catalogItem.useOwnClock) {
+            // This is probably not needed, since we should be doing it on update, but just do this here explicitly
+            // to be sure it is immediately current before we change.
+            updateCurrentTime(catalogItem, catalogItem.terria.clock.currentTime);
+        } else {
+            catalogItem.terria.clock.currentTime = JulianDate.clone(catalogItem.clock.currentTime);
+        }
+    }
+
+    catalogItem._lastUseOwnClock = catalogItem.useOwnClock;
+}
+
+function timeAtIndex(catalogItem, index) {
+    if (defined(index) && defined(catalogItem.availableDates) && (index >= 0) && (index < catalogItem.availableDates.length)) {
+        return catalogItem.availableDates[index];
+    }
+
+    return undefined;
+}
+
+function getCurrentIndex(catalogItem) {
+    return catalogItem.intervals.indexOf(catalogItem.currentTime);
+}
+
+function updateIndex (catalogItem, index) {
+    const time = timeAtIndex(catalogItem, index);
+
+    if (defined(time)) {
+        catalogItem.useOwnClock = true;
+        updateCurrentTime(catalogItem, JulianDate.fromDate(new Date(time)));
+        catalogItem.useClock();
+    }
+}
+
+function nextIndex(catalogItem) {
+    const index = getCurrentIndex(catalogItem);
+    if(defined(index) && (index < (catalogItem.intervals.length-1))) {
```

Comment:
more nits: missing space after if, and all the parenthesis are unnecessary except the outer ones.

Reply:
Done 19cc3c1f3cbd09c81f2de94c34a560c6c7bbde87

Note: I only removed one set of parenthesis since I consider the others reasonable practice since it somewhat stops people falling victim to wondering about operator precedents and makes the intention explicit.